bson/_cbsonmodule.c

```diff
@@ -2797,6 +2816,85 @@ static PyObject* _cbson_decode_all(PyObject* self, PyObject* args) {
     return result;
 }
 
+
+static PyObject* _cbson_array_of_documents_to_buffer(PyObject* self, PyObject* args) {
+    uint32_t size;
+    uint32_t value_length;
+    uint32_t position = 0;
+    buffer_t buffer;
+    const char* string;
+    PyObject* arr;
+    PyObject* result = NULL;
+    Py_buffer view = {0};
+
+    if (!PyArg_ParseTuple(args, "O", &arr)) {
+        return NULL;
+    }
+
+    buffer = pymongo_buffer_new();
+
+    if (!_get_buffer(arr, &view)) {
+        goto fail;
+    }
+
+    string = (char*)view.buf;
+    memcpy(&size, string, 4);
+    size = BSON_UINT32_FROM_LE(size);
+
+    if (view.len < size || view.len > BSON_MAX_SIZE) {
+        PyObject* InvalidBSON = _error("InvalidBSON");
+        if (InvalidBSON) {
+            PyErr_SetString(InvalidBSON, "objsize too large");
+            Py_DECREF(InvalidBSON);
+        }
+        goto done;
+    }
+
+    position += 4;
+    while (position < size - 1) {
+        // Just skip the keys.
+        while (string[position] != '\0') {
+            position++;
+        }
+        position++;
+
+        memcpy(&value_length, string + position, 4);
+        value_length = BSON_UINT32_FROM_LE(value_length);
```

Comment:
Do we need to validate value_length? Perhaps we should add a _pymongo_get_object_size helper method?

Reply:
Added validation

Reply:
This validation is different from the other places we parse obj/arr sizes. Looks like it was accidentally copied from bson string decoding. See decode_all for the correct validation (including the necessary InvalidBSON errors): https://github.com/mongodb/mongo-python-driver/blob/e07081599301fe9e53387f98ccc00d7d2c700eb0/bson/_cbsonmodule.c#L2743-L2773

Reply:
Done